Area and band plots

senier opened the pull request:
This PR implements area and band plots. In addition I did some refactoring and documentation of `plot_graph` as it got even a bit more complex.

I changed the Short-/Long-term load graph to use the band plot and the set volume graphs to use area plots. We should discuss what else we want to change to make graphs consistent. My current thinking was that the area graph is most appropriate in graphs that represent "an area under the curve" like loads. But of course we could use them as a style element elsewhere (e.g. use area plots for averages and dots for single values, e.g. in body weight). We should try some options here.

The yellow style of the Short-/Long-term band plot is terrible on both, light and dark background IMHO. This should definitely be improved.

senier commented:
> That's true. I think this would be a nice alternative:
> 
> ![image](https://private-user-images.githubusercontent.com/15232394/395341109-0ec8156e-238a-4ab4-9779-621db4f54944.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQxOTc0NjUsIm5iZiI6MTczNDE5NzE2NSwicGF0aCI6Ii8xNTIzMjM5NC8zOTUzNDExMDktMGVjODE1NmUtMjM4YS00YWI0LTk3NzktNjIxZGI0ZjU0OTQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE0VDE3MjYwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIyOWViYmM2NjkzMWY5NTUzMjJmOTVjNjExMjEwNmNlZmE0YTdiNTc2YzQxZGNmZWY0YmQ4NmI5ODQwN2IzZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.zs3U-Qx1CKNVkI1NYJkQhNObrmZul2NqWrV-wXoDNeI)

I agree that that's better. The only issue that I see is that it shows the same color for short-term and long-term load in the legend. This might be confusing:

![image](https://github.com/user-attachments/assets/429c095a-45bb-48e4-a007-741f56eae77c)

Here is another idea which I find quite okay:

![image](https://github.com/user-attachments/assets/fa7c2cd8-4e0e-442c-8851-7d6b2f82d0e6)
![image](https://github.com/user-attachments/assets/f5757de7-5190-4780-8f3a-48e4b560a089)

senier commented:
Some more ideas for the exercise page:

- Use area plots for set volume and volume load
- Use a band plot for RIR (I don't love the color, but orange was worse for that context IMO)
- Remove dots from weight

![image](https://github.com/user-attachments/assets/2c041d57-d5eb-4d6f-8f70-87c15b0b2a9b)
![image](https://github.com/user-attachments/assets/d0989983-1772-4d85-913d-aa67d874507c)

senier commented:
Another idea for the weight chart:

![image](https://github.com/user-attachments/assets/32c5a9e4-359c-4ffe-83bb-4b1feb3a7ddb)

senier commented:
Here is an idea for a general approach that we might consider adopting. For volumes (where "area under the curve" is relevant) use area plots with *the same* color as line. For plain values or averages use are plots with gray areas and a colored line. While this will only work nicely for single value graphs (or stacked graph as Reps+RIR where values have the same kind), it makes those graphs visually more consistent:

![image](https://github.com/user-attachments/assets/b6d0e8b1-243a-4612-b4a5-196e1b8b1bbf)
![image](https://github.com/user-attachments/assets/4c645a80-f656-4d25-aa09-f55d09d04d99)

![image](https://github.com/user-attachments/assets/9bf1e287-ff65-4f1a-84c6-b216161c91d8)
![image](https://github.com/user-attachments/assets/065d45de-cb82-46f5-9486-17eac0e9c207)

treiher commented:
> I agree that that's better. The only issue that I see is that it shows the same color for short-term and long-term load in the legend. This might be confusing:

I think we should use the same color in the legend as is actually used in the band plot:

![image](https://github.com/user-attachments/assets/fcf19417-17b8-4293-8ced-902c03d6a19f)

![image](https://github.com/user-attachments/assets/01937c12-5e5c-44df-b8b3-540cb92a61fe)

Given that, I don't see any reason why we shouldn't use the green one. The grey one looks okay, but I like the green one better.

> Some more ideas for the exercise page:
> 
> * Use area plots for set volume and volume load

Looks good to me.

> * Use a band plot for RIR (I don't love the color, but orange was worse for that context IMO)

I like the idea! I'm just not conviced of the colors. Using the same color just with less opacity looks better in my opinion:

![image](https://github.com/user-attachments/assets/59fe1e27-bdfb-4e2f-93f3-9ad7a36b28c6)

> * Remove dots from weight

Removing the dots makes sense for consistency.

> Another idea for the weight chart:
> 
> ![image](https://private-user-images.githubusercontent.com/6357197/395821087-32c5a9e4-359c-4ffe-83bb-4b1feb3a7ddb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQyNTQ2OTYsIm5iZiI6MTczNDI1NDM5NiwicGF0aCI6Ii82MzU3MTk3LzM5NTgyMTA4Ny0zMmM1YTllNC0zNTljLTRmZmUtODNiYi00YjFmZWIzYTdkZGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTVUMDkxOTU2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWQ1ZTkxMjhkY2UzNTM2YWQwNTEwYzhlYTg1ZDA3Yzc3NTZlNTU3NjdkZjA0MThjMDliZTlhMzk4MTZlMTUzOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.RiTmW3F0AnJrbNz6utqGWqW3nE7qnWNYpD3DNL3sMCA) ![image](https://private-user-images.githubusercontent.com/6357197/395821102-c43559b4-f80e-4108-8440-1a3f633d0a40.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQyNTQ2OTYsIm5iZiI6MTczNDI1NDM5NiwicGF0aCI6Ii82MzU3MTk3LzM5NTgyMTEwMi1jNDM1NTliNC1mODBlLTQxMDgtODQ0MC0xYTNmNjMzZDBhNDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTVUMDkxOTU2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjFkODFkOTc4YTJhZjFhZGI2MWYyMmViOTg4NThlYThhODUyODE3NmVjNTU0Yzc4Mjk5N2ZhZjNiMTgwZDQwZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.lXsmVrka1A5P3b4gHkKg3GWjEDa5XgJFolI8pLolzoA)

In this chart, the actual values are more prominent than the average. I think it should be the other way round. Simply switching both styles might be better, but I think it would still not be great.

Here are some other ideas:

We could use a band plot to visualize the variance of the actual values from the average:

![image](https://github.com/user-attachments/assets/ed9b3306-b72c-4d34-8cd8-634ed4118e11)

Or use a line plot with less opacity:

![image](https://github.com/user-attachments/assets/aacde590-65dd-410e-97a0-99604252359f)

The band plot looks nicer to me.

> Here is an idea for a general approach that we might consider adopting. For volumes (where "area under the curve" is relevant) use area plots with **the same** color as the line. For plain values or averages use plots with gray areas and a colored line. While this will only work nicely for single value graphs (or stacked graph as Reps+RIR where values have the same kind), it makes those graphs visually more consistent:

The combination of a colored line and a grey area plot doesn't look so nice to me (especially in the combination with the light green and red line). I would rather omit the grey area plot for averages.

I think it would be also nice to use a band plot to show the variance for averages, similar to the body weight chart shown above. Here is an example for the weight of an exercise. There is a variance due to drop sets:

![image](https://github.com/user-attachments/assets/af02b6f2-2c9a-4c27-849f-3e56d8414ad6)

I would suggest the following general approach:

- For volumes, use area plots of the same color as the line.
- For averages use line plots. For plain values or other additional information, use band plots of the same color as the average but with less opacity.

senier commented:
I pushed my current WIP (this will need cleanup later).

> Given that, I don't see any reason why we shouldn't use the green one. The grey one looks okay, but I like the green one better.

I implemented that (and fixed the label coloring to support alpha values).

> I like the idea! I'm just not conviced of the colors. Using the same color just with less opacity looks better in my opinion:

That's nice. I implemented it that way.

> We could use a band plot to visualize the variance of the actual values from the average:
> [...]
> The band plot looks nicer to me.

I agree - that's a great idea and how I implemented it.

> I think it would be also nice to use a band plot to show the variance for averages, similar to the body weight chart shown above. Here is an example for the weight of an exercise. There is a variance due to drop sets:

I implemented this, please have a look (my graph looks much different due to DUP).

> * For volumes, use area plots of the same color as the line.
> * For averages use line plots. For plain values or other additional information, use band plots of the same color as the average but with less opacity.

Sounds good. Maybe you want to add this to the wiki?

So far there are two open questions:

1. Are the TUT charts considered volume-like? I'm using line plots for now, but this is easily changed.
2. The body fat chart uses lines now. I'm not totally satisfied with it, yet, as it can look pretty chaotic. I don't have a good idea here, though.

treiher commented:
> I implemented this, please have a look (my graph looks much different due to DUP).

I think the current implementation just shows the minimum and maximum of the average weight (of all trainings sessions at the same day), whereas it should show the minimum and maximum of the actual weight. This would explain the flat line of 26.6 kg when the actual weight varied from 20 kg to 35 kg in the training session:

![image](https://github.com/user-attachments/assets/59fdd96c-bb67-44ff-884a-6f15e34395fa)

I'm not sure why there's a tiny band plot visible when the average weight changes. I would expect there to be no band plot if only one weight was used each day, but this is not the case.

> > * For volumes, use area plots of the same color as the line.
> > * For averages use line plots. For plain values or other additional information, use band plots of the same color as the average but with less opacity.
> 
> Sounds good. Maybe you want to add this to the wiki?

Good idea! I will do that.

> So far there are two open questions:
> 
> 1. Are the TUT charts considered volume-like? I'm using line plots for now, but this is easily changed.

Yes, I think so. TUT also represents a total. So it would make sense to use the same style as for the volume plots.

Also, it would make sense to represent time and RPE the same as weight. These are all averages.

> 2. The body fat chart uses lines now. I'm not totally satisfied with it, yet, as it can look pretty chaotic. I don't have a good idea here, though.

I also think that this chart is not as nice as the others.

![image](https://github.com/user-attachments/assets/9c54a634-e0ff-4bbe-9f17-f0fa4684616b)

We should use the same style for the body weight as in the body weight chart where the average is more prominent:

![image](https://github.com/user-attachments/assets/ee4513f8-50bf-4aad-bfb0-1f2eec1cba81)

Split the chart into two: One for JP3 and one for JP7. And use the same color, but with different opacity:

![image](https://github.com/user-attachments/assets/3bfac220-a02b-4606-9c02-ab6a1e7c9293)

senier commented:
> I'm not sure why there's a tiny band plot visible when the average weight changes. I would expect there to be no band plot if only one weight was used each day, but this is not the case.

I think I got the idea wrong. This should be fixed now - please have another look.

> Yes, I think so. TUT also represents a total. So it would make sense to use the same style as for the volume plots.

This is done.

> Also, it would make sense to represent time and RPE the same as weight. These are all averages.

This should be the case now. Please have a look as I don't have good data for those charts.

> We should use the same style for the body weight as in the body weight chart where the average is more prominent:
> 
> An option would be to split the chart into two (one for JP3 and one for JP7) and use the same color, but with different opacity:

Splitting the graphs is a great idea! I kept the color for the JP3/7 charts though, as we already have two levels of opacity for the weight chart and I found this much less readable.

treiher commented:
The new style looks really nice! I don't see any errors in the charts. I'm just wondering if we should change the color of the repetitions chart (e.g., to orange), so that we have a distinct color for all aspects in the training charts. Blue is the only color that is used twice.

senier commented:
> The new style looks really nice!

I agree - can't wait to use it :smile:

> I don't see any errors in the charts. I'm just wondering if we should change the color of the repetitions chart (e.g., to orange), so that we have a distinct color for all aspects in the training charts. Blue is the only color that is used twice.

That makes sense - I changed it to orange which IMHO works well.

The PR is now ready.

senier commented on frontend/src/ui/page/training.rs:
```diff
@@ -529,63 +529,64 @@ pub fn view_charts<Ms>(
+                        plots: common::plot_area_with_border(
+                            common::COLOR_LONG_TERM_LOAD,
+                            common::COLOR_LONG_TERM_LOAD,
+                            0.2,
```
I introduced constants for line/area opacity as well as line width.

treiher replied:
So far the constants are only used in the plotting functions. It would make sense to use the opacity constants in the legend as well.

senier replied:
That's true. Changed.

senier commented on frontend/src/ui/page/exercise.rs:
```diff
@@ -477,142 +475,158 @@ pub fn view_charts<Ms>(
-                    &[common::PlotData{
-                        values: time.into_iter()
-                            .map(|(date, values)| {
+                &[("Time (s)", common::COLOR_TIME, 0.9)],
```
Done.

senier commented on frontend/src/ui/common.rs:
```diff
@@ -32,20 +31,34 @@ pub const FONT: (&str, u32) = ("Roboto", 11);
+        PlotType::Area(area_color, opacity),
+        PlotType::Line(line_color, 0.9, width),
+    ]
+}
```
Those are meant to be convenience functions anyways (you can always use the `PlotType` primitives to do more complex things). I removed those configuration options (including the line width in `plot_area_with_border` which also never changed).

senier commented on frontend/src/ui/page/exercise.rs:
```diff
@@ -477,142 +475,158 @@ pub fn view_charts<Ms>(
+                    common::COLOR_REPS_RIR,
+                    0.3,
+                    0,
+                ),
```
Changed.
